src/djangooidc/views.py

```diff
@@ -108,16 +108,25 @@ def login_callback(request):
         if user:
             login(request, user)
             logger.info("Successfully logged in user %s" % user)
-            # Double login bug (1507)?
+            # Clear the flag if the exception is not caught
+            request.session.pop("redirect_attempted", None)
             return redirect(request.session.get("next", "/"))
         else:
             raise o_e.BannedUser()
-    except o_e.NoStateDefined as nsd_err:
-        # In the event that a user is in the middle of a login when the app is restarted,
-        # their session state will no longer be available, so redirect the user to the
-        # beginning of login process without raising an error to the user.
-        logger.warning(f"No State Defined: {nsd_err}")
-        return redirect(request.session.get("next", "/"))
+    except o_e.StateMismatch as nsd_err:
+        # Check if the redirect has already been attempted
+        if not request.session.get("redirect_attempted", False):
+            # Set the flag to indicate that the redirect has been attempted
+            request.session["redirect_attempted"] = True
+
+            # In the event of a state mismatch between OP and session, redirect the user to the
+            # beginning of login process without raising an error to the user. Attempt once.
+            logger.warning(f"No State Defined: {nsd_err}")
+            return redirect(request.session.get("next", "/"))
+        else:
+            # Clear the flag if the exception is not caught
+            request.session.pop("redirect_attempted", None)
+            return error_page(request, nsd_err)
```

Comment:
**blocking question**

if I am reading this right, error_page(request, nsd_err) will call the error_page function but with error equal to the StateMismatch error, but in error_page it's checking if error is an instance of AuthenticationFailed or InternalError and then for all others it's throwing a 500. 

It looks like with this flow if a user ever gets to this line they will then see a 500, is that what we want?

I would think we want to 1.) just have StateMismatch inherit from AuthenticationFailed or 2.) also check for "is instance StateMatch) in error_page. 

Thoughts?

Reply:
Going with 1